pkg/controller/vitessshard/reconcile_disk.go

```diff
@@ -0,0 +1,126 @@
+package vitessshard
+
+import (
+	"context"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	planetscalev2 "planetscale.dev/vitess-operator/pkg/apis/planetscale/v2"
+	"planetscale.dev/vitess-operator/pkg/operator/results"
+	"planetscale.dev/vitess-operator/pkg/operator/rollout"
+)
+
+const (
+	pvcDiskSizeAnnotation = "pvc-disk-size"
+)
+
+func (r *ReconcileVitessShard) reconcileDisk(ctx context.Context, vts *planetscalev2.VitessShard) (reconcile.Result, error) {
+	resultBuilder := &results.Builder{}
+	changed := false
+	tabletKeys := tabletKeysFromShard(vts)
+	tabletPods, err := r.tabletPodsFromShard(ctx, vts)
+	if err != nil {
+		return resultBuilder.Error(err)
+	}
+
+	for i  := range vts.Spec.TabletPools {
+		tabletPool := &vts.Spec.TabletPools[i]
+		requestedDiskQuantity := tabletPool.DataVolumeClaimTemplate.Resources.Requests[v1.ResourceStorage]
+		requestedDisk := requestedDiskQuantity.String()
+		poolTablets := tabletsForPool(vts, tabletKeys, tabletPool.Cell, string(tabletPool.Type))
+
+		for _, tabletKey := range poolTablets {
+			pod, ok := tabletPods[tabletKey]
+			if !ok {
+				continue
+			}
+
+			tabletDiskSize, ok := pod.Annotations[pvcDiskSizeAnnotation]
+			if !ok {
+				// If it's never been set before, apply the disk size annotation.
+				setDiskSizeAnnotation(pod, requestedDisk)
+				continue
+			}
+
+			// If disk size is unchanged, move on to the next tablet.
+			if tabletDiskSize == requestedDisk {
+				continue
+			}
+
+			// If there are disk size changes, mark the shard as changed to cascade later.
+			changed = true
```

Comment:
This would begin cascading as soon as *any* tablet is ready. We need to wait until *all* tablets (that need restarts) are ready.